File: cpp/187.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iterator>
#include <utility>
#include <map>
#include <set>
#include <string>
#include <deque>
#include <unordered_map>
#include <unordered_set>
#include <climits>
#include <numeric>
#include <sstream>
using namespace std;

template<typename T>
void prints(vector<T> &v){
  cout << "[";
  for(uint i=0; i<v.size()-1; ++i)
    cout << v[i] << ", ";
  if(v.size()) cout << v[v.size()-1];
  cout << "]\n";
}

class Solution {
public:
  int hash(string &s, int st){
    int ans = 0, t = 1;
    for(int i=9; i>=0; --i){
      int c;
      switch(s[st + i]){
      case 'A': c = 0; break;
      case 'C': c = 1; break;
      case 'G': c = 2; break;
      case 'T': c = 3; break;
      }
      ans += c * t;
      t <<= 2;
    }
    return ans;
  }
  vector<string> findRepeatedDnaSequences(string s) {
    int m[1050000];
    for(int i=0; i<1050000; i++)
      m[i] = 0;
    vector<string> ans;
    int n = s.length();
    for(int i=0; i+10<=n; i++){
      int t = hash(s, i);
      ++m[t];
      if(m[t] == 2)
        ans.push_back(s.substr(i, 10));
    }
    return ans;
  }
};


int main(){
  auto sol = Solution();
}

File: python/187.py
class Solution(object):
    def findRepeatedDnaSequences(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        n = len(s)
        if n < 10:
            return []
        a = [0] * 1050000
        ss = s
        s = [(0 if c == 'A' else 1 if c == 'C' else 2 if c == 'G' else 3) for c in s]
        h = 0
        for i in xrange(10):
            h += s[i] * (1<<2*i)
        a[h] = 1
        ans = []
        for i in xrange(1, n-10+1):
            h >>= 2
            h +=  s[i+9] * (1<<18)
            if a[h] == 1:
                ans.append(ss[i:i+10])
            a[h] += 1
        return ans
